joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/datastructures/ProgramSummary.scala

```diff
@@ -0,0 +1,253 @@
+    *   the shorthand name.
+    * @return
+    *   the type meta-data if found.
+    */
+  def tryResolveTypeReference(typeName: String): Option[T] = {
+    typesInScope
+      .collectFirst {
+        case typ if typ.name.endsWith(typeName)                                           => typ
+        case typ if aliasedTypes.contains(typeName) && typ.name == aliasedTypes(typeName) => typ
+      }
+      .flatMap(typ => summary.namespaceFor(typ).map(namespace => typ))
+  }
+
+  /** Given the type full name and call name, will attempt to find the matching entry.
+    *
+    * @param typeFullName
+    *   the base type full name.
+    * @param callName
+    *   the call name.
+    * @param argTypes
+    *   the observed argument types. Only relevant for languages that implement overloading.
+    * @return
+    *   the method full name if found.
+    */
+  def tryResolveMethodInvocation(typeFullName: String, callName: String, argTypes: List[String]): Option[M] = {
+    tryResolveTypeReference(typeFullName).flatMap { t =>
+      t.methods.find { m => m.name == callName && isOverloadedBy(m, argTypes) }
+    }
+  }
+
+  /** Determines if, by observing the given argument types, that the method's signature is a plausible match to the
+    * observed arguments.
+    *
+    * The default implementation only considers that the same number of arguments are added and does not account for
+    * variadic arguments nor polymorphism.
+    *
+    * @param method
+    *   the method meta data.
+    * @param argTypes
+    *   the observed arguments from the call-site.
+    * @return
+    *   true if the method could be overloaded by a call with these argument types.
+    */
+  protected def isOverloadedBy(method: M, argTypes: List[String]): Boolean = {
+    method.parameterTypes.size == argTypes.size
+  }
+
+  /** Appends known types imported into the scope.
+    * @param namespace
+    *   the fully qualified imported namespace.
+    */
+  def addImportedNamespace(namespace: String): Unit = {
+    val knownTypesFromNamespace = summary.typesUnderNamespace(namespace)
+    typesInScope.addAll(knownTypesFromNamespace)
+  }
+
+  /** Appends known types imported into the scope.
+    * @param typeOrModule
+    *   the type name or full name.
+    */
+  def addImportedTypeOrModule(typeOrModule: String): Unit = {
+    val matchingTypes = summary.matchingTypes(typeOrModule)
+    typesInScope.addAll(matchingTypes)
+  }
+
+}
+
+/** An implementation of combining the typed scoping structures to manage the available type information at namespace
+  * levels.
+  *
+  * @param summary
+  *   the program summary.
+  */
+class DefaultTypedScope[I, V, S <: TypedScopeElement, M <: MethodLike, F <: FieldLike, T <: TypeLike[M, F]](
+  summary: ProgramSummary[M, F, T]
+) extends Scope[I, V, S]
+    with TypedScope[M, F, T, S](summary) {
+
+  /** Pops the scope, adding types from the scope if necessary.
+    */
+  override def pushNewScope(scopeNode: S): Unit = {
+    scopeNode match
```

Comment:
Minor note: I'd prefer to keep our current style to use braces around match statements - we didn't jump onto the bandwagon of significant indentation, and IMO that's a good thing.